### PR TITLE
Make `IoVecBuf` traits more generic

### PR DESCRIPTION
**Pull request by jon-chuang**
related: https://github.com/bytedance/monoio/issues/53

### CODE REVIEW

File: monoio/src/buf/io_vec_buf.rs [hunk 1]
```diff
@@ -175,11 +175,11 @@ unsafe impl IoVecBufMut for VecBuf {
         for (idx, iovec) in self.iovecs.iter_mut().enumerate() {
             if iovec.iov_len <= len {
                 // set_init all
-                self.raw[idx].set_len(iovec.iov_len);
+                self.raw[idx].set_init(iovec.iov_len);
```
Comment (jon-chuang) [inline monoio/src/buf/io_vec_buf.rs:178]: Not sure on this behaviour. Not all IoBufMut mutate len for `set_init`

Comment (jon-chuang) [reply monoio/src/buf/io_vec_buf.rs:178]: Notably, only Vec<u8> and bytes::BytesMut are mutated

File: monoio/src/buf/io_vec_buf.rs [hunk 2]
```diff
@@ -162,7 +162,7 @@ pub unsafe trait IoVecBufMut: Unpin + 'static {
     unsafe fn set_init(&mut self, pos: usize);
 }
 
-unsafe impl IoVecBufMut for VecBuf {
+unsafe impl<B: IoBufMut + IoBuf> IoVecBufMut for VecBuf<B> {
```
Comment (jon-chuang) [inline monoio/src/buf/io_vec_buf.rs:165]: here we are assuming for simplicity that all `B: IoBufMut` also has `B: IoBuf`. If not, we can also decouple.

Comment (jon-chuang) [inline monoio/src/buf/io_vec_buf.rs:165]: here we are assuming for simplicity that all B: IoBufMut also has B: IoBuf. 

If not, we can also decouple if required.

File: monoio/src/buf/io_vec_buf.rs (hunk 1 above)
Comment (jon-chuang) [inline monoio/src/buf/io_vec_buf.rs:178]: Not sure on this behaviour. Not all `IoBufMut` mutate buffer for `set_init`
Notably, only `Vec<u8>` and `bytes::BytesMut` are mutated by `IoBufMut::set_init`

File: monoio/src/buf/io_vec_buf.rs [hunk 3]
```diff
@@ -175,11 +175,11 @@ unsafe impl IoVecBufMut for VecBuf {
         for (idx, iovec) in self.iovecs.iter_mut().enumerate() {
             if iovec.iov_len <= len {
                 // set_init all
-                self.raw[idx].set_len(iovec.iov_len);
+                self.raw[idx].set_init(iovec.iov_len);
                 len -= iovec.iov_len;
             } else {
                 if len > 0 {
```
Comment (jon-chuang) [inline monoio/src/buf/io_vec_buf.rs:181]: Shouldn't we set to 0 even if len is 0?

File: monoio/src/buf/io_vec_buf.rs (hunk 2 above)
Comment (dyxushuai) [reply monoio/src/buf/io_vec_buf.rs:165]: Yes, we have split `mut` and `ref`  buf trait without ambiguity.

File: monoio/src/buf/io_vec_buf.rs (hunk 1 above)
Comment (dyxushuai) [reply monoio/src/buf/io_vec_buf.rs:178]: We need write back the length from uring&kernel. So all `BufMut` types must be able to set length.
So `unsafe` is necessary!

File: monoio/src/buf/io_vec_buf.rs (hunk 3 above)
Comment (dyxushuai) [reply monoio/src/buf/io_vec_buf.rs:181]: IO ops from kernel and return zero length means EOF

**Comment by ihciah** [comment]
Sorry for late checking it because this PR not merged for very long and I used to ignore it...

Do you still have interesting in merging it? I think it is more generic to make bytes buffer as a B, we can use something like Bytes with this ability. But to keep it compatiable, maybe adding a default type Vec<u8> to B is better.

Sorry for no responding again. If you are interested, please reopen it or open another PR, thanks!

**Pull request closed without merging**